packages/core/src/browser/tree/tree-model.ts

```diff
@@ -158,53 +161,60 @@ export class TreeModel implements ITreeModel, SelectionProvider<Readonly<ISelect
         }
     }
 
-    get selectedNode() {
-        return this.selection.selectedNode;
+    get selectedNodes() {
+        return this.selection.selectedNodes;
     }
 
     get onSelectionChanged() {
         return this.selection.onSelectionChanged;
     }
 
-    selectNode(node: ISelectableTreeNode | undefined): void {
-        this.selection.selectNode(node);
+    selectNode(node: ISelectableTreeNode, options?: ITreeSelectionService.Options): void {
```

Comment:
Should not be multi-selection support configured per a tree model?

Reply:
No. It is per tree widget. The model itself will be always multi-select-aware. We need to propagate the key mask from the widget to the model, so I have changed it to an enumeration:`SINGLE`, `MULTI_INDIVIDUAL` (<kbd>Ctrl/Cmd</kbd>), and `MULTI_RANGE` (<kbd>Shift</kbd>).

Reply:
But do not get me wrong, you're right. The `options` argument naming is misleading, so I will change it to `selectionType`.

Reply:
It would be provide 2-level of APIs.
1. minimal to support any use case
```ts
setSelection(selection: ITreeNode[]): void
getSelection(): ITreeNode[]
readonly onDidSelectionChanged: Event<ITreeNode[]>();
```
2. convenient API per a use case which should be implemented with a minmial api, e.g.
- `toggleSelection(node)` - to toggle a selection of one node
- `selectNode(node)` - to add a node to a selection
- `deselectNode(node)` - to add remove a node from a selection
- `selectRange(node, node)` - to a select a range of nodes between two
- `selectRange(node)` - to a select a range of nodes between given and last selected
- `deselectRange`
- ...

Reply:
I agree with the `toggleNode`. That is useful. Not with the others.

Reply:
I concern more about having the minimal layer that a user can implement whatever he wants. It is a typical technique: https://martinfowler.com/bliki/MinimalInterface.html. It is how the tree model was designed from the beginning.

I don't think that forcing `selectNode` do all kind of selection functionality depending on the given options is a good idea. Having set of small methods which get selected nodes, mutate them and set for different cases is easier to understand and maintain.

Reply:
OK, thanks! I will do like that.

Reply:
> `selectNode(node)`

This will discard the previous selection state. In some cases, I need to keep the previous state and append/remove an item, so I either need a dedicated method `addToSelection(node)` or have a `selectNode(node: STN, merge?: boolean)`. What do you think, @akosyakov.

Reply:
Something like this:
```typescript

    /**
     * Expands the given node. If the `node` argument is `undefined`, then expands the currently selected tree node.
     * If multiple tree nodes are selected, expands the most recently selected tree node.
     */
    expandNode(node?: Readonly<ExpandableTreeNode>): boolean;

    /**
     * Collapses the given node. If the `node` argument is `undefined`, then collapses the currently selected tree node.
     * If multiple tree nodes are selected, collapses the most recently selected tree node.
     */
    collapseNode(node?: Readonly<ExpandableTreeNode>): boolean;

    /**
     * Toggles the expansion state of the given node. If not give, then it toggles the expansion state of the currently selected node.
     * If multiple nodes are selected, then the most recently selected tree node's expansion state will be toggled.
     */
    toggleNodeExpansion(node?: Readonly<ExpandableTreeNode>): void;

    /**
     * Opens the given node or the currently selected on if the argument is `undefined`.
     * If multiple nodes are selected, open the most recently selected node.
     */
    openNode(node?: Readonly<TreeNode> | undefined): void;

    /**
     * Event when a node should be opened.
     */
    readonly onOpenNode: Event<Readonly<TreeNode>>;

    /**
     * Selects the parent node relatively to the selected taking into account node expansion.
     */
    selectParent(): void;

    /**
     * Navigates to the given node if it is defined.
     * Navigation sets a node as a root node and expand it.
     */
    navigateTo(node: Readonly<TreeNode> | undefined): void;

    /**
     * Tests whether it is possible to navigate forward.
     */
    canNavigateForward(): boolean;

    /**
     * Tests whether it is possible to navigate backward.
     */
    canNavigateBackward(): boolean;

    /**
     * Navigates forward.
     */
    navigateForward(): void;

    /**
     * Navigates backward.
     */
    navigateBackward(): void;

    /**
     * Selects the previous node relatively to the currently selected one. This method takes the expansion state of the tree into consideration.
     */
    selectPrevNode(keepState?: boolean): void;

    /**
     * Selects the next node relatively to the currently selected one. This method takes the expansion state of the tree into consideration.
     */
    selectNextNode(keepState?: boolean): void;

    /**
     * Selects the tree node argument.
     *
     *  - If the `keepState` argument is `false` or `undefined`, then it has the same effect as invoking `setSelection([node])`. The previous selection
     * state will be overridden with given single tree node.
     *  - If `keepState` is `true`, then the previous selection state will be kept and the `node` argument will be merged into the previous selection state.
     *  - If `keepState` is `true` and the `node` was already selected, the selection content will not change but the order of the selected items. The
     * `node` argument will be the most recently selected item.
     */
    selectNode(node: Readonly<SelectableTreeNode>, keepState?: boolean): void;

    /**
     * Unselects the given tree node argument. Has no effect, if the argument is not a valid tree node, or was not yet selected.
     */
    unselectNode(node: Readonly<SelectableTreeNode>): void;

    /**
     * Toggles the selection state of the `node` argument. If the node was not yet selected, it will be. Toggling the state of the tree node
     * will keep any previous selection state. If the `node` was selected, `unselects` the node.
     */
    toggleSelection(node: Readonly<SelectableTreeNode>): void;

    /**
     * Calculates and sets the selection range between the `fromNode` and the `toNode`.
     *  - Has not effect if either the `fromNode` or the `toNode` does not exist or invalid.
     *  - If `keepState` is `true`, then the range of selection will be merged into the previous selection state.
     *  - If `keepState` is `false` or `undefined`, then the previous state will be overridden with the selection range.
     *  - If `fromNode` is `undefined`, then the most recently selected tree node will be used instead.
     */
    selectRange(toNode: Readonly<SelectableTreeNode>, fromNode?: Readonly<SelectableTreeNode>, keepState?: boolean): void;

    /**
     * Counterpart of the `selectRange` method. Calculates and removes the selection of the nodes which contained in the range.
     */
    unselectRange(toNode: Readonly<SelectableTreeNode>, fromNode?: Readonly<SelectableTreeNode>): void;
```

Reply:
sounds good, maybe name `keepSelection` instead `keepState`?